### ISSUE TITLE
storage/sinks: in-place schema changes (`ALTER SINK`)

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#26350]
## Overview

Allow changing a sink's underlying relation with `ALTER SINK ... FROM <new_rel>`.

Otherwise known as `ALTER SINK` or "sink schema changes."

## Details

[Design doc](https://www.notion.so/materialize/ALTER-SINK-MVP-ffa7e969830a49eb979043510425e3b0)

Imagine there was a sink built on `foo` that is emitting the "important" records from foo:

```
CREATE MATERIALIZED VIEW foo_important AS SELECT a FROM foo WHERE a % 3 = 0;
CREATE SINK foo_snk FROM foo_important INTO KAFKA ...;
```

We could introduce an `ALTER SINK` command that allows cutting a sink over to a new input:

```
CREATE MATERIALIZED VIEW foo_important_new AS SELECT a FROM foo WHERE a % 3 = 0 AND length(b) > 5;
ALTER SINK foo_snk FROM foo_important_new;
```

The `ALTER SINK` command would pick some timestamp at which to start emitting records from `foo_important_new` instead of `foo_new`. In the example above, the sink would pick a consistent time at which to start using the new definition of "importance" that takes into account the new field in the input data. This doesn't always yield sensible results—the new input might retract a record that exists in its timeline, but not in the sink's timeline—but many downstream clients in the Kafka ecosystem can tolerate this sort of inconsistency; e.g., we've heard repeatedly "can you just not show me retractions?" as a feature request.

We could even allow `ALTER SINK` to cut over to an input of a different shape. This changes the output shape, but there are many shape evolutions that are considered backwards compatible by Avro.

The approach becomes quite a bit more ergonomic with dbt in the mix. If you manage your sources, views, and sinks with dbt, rewriting the whole stack to use new sources, new views based off those sources, and then updating the sinks to point at those new views is something that our adapter could handle automatically.

Pairing this with #17021 would allow users to do an end-to-end schema change, from their upstream systems, through Materialize, and into their downstream systems.

## Related work

* #26951

## See also

* #15897
* #17021

cc @josharenberg @morsapaes

## Action log

* **5 Jun 2023** Implemented by [adapter: implement alter sink #27379](https://github.com/MaterializeInc/materialize/pull/27379)
* **7 May 2024.** Assigned to @petrosagg, who will start by scoping the work and in particular considering the relationship to https://github.com/MaterializeInc/materialize/issues/23508.

**Comment by bosconi** [comment]
Implemented by https://github.com/MaterializeInc/materialize/pull/27379

**Comment by bosconi** [comment]
@petrosagg what if anything is left to do before we roll this out into PrPr? (https://github.com/MaterializeInc/materialize/issues/27501 was put to bed, it looks like.)

I forget if there's a feature flag. If so, will we keep it off for now and enable on case by case basis? 

Also, is there obvious follow up work? If so, should we create an epic to track it?

cc @morsapaes

**Comment by petrosagg** [comment]
recording here my answers in slack:
* We don't have a feature flag
* There was a failing test in nightly but it turned out to be just a technicality with the test interacting badly with redpanda
* I'm not sure we have any follow up work except for writing documentation

**Comment by petrosagg** [comment]
I guess one follow up work item is to detect incompatible schema changes during the `ALTER SINK` planning as opposed to after the fact. Currently the user will end up with a faulty sink that can't be salvaged if they accidentally alter it and point it to an incompatible relation

**Comment by benesch** [comment]
> I guess one follow up work item is to detect incompatible schema changes during the `ALTER SINK` planning as opposed to after the fact. Currently the user will end up with a faulty sink that can't be salvaged if they accidentally alter it and point it to an incompatible relation

I pitched @rjobanp on that here: https://github.com/MaterializeInc/materialize/issues/23794#issuecomment-2162049350

**Comment by morsapaes** [comment]
Fixed in #27379, documented in #27594. Opened #27613 to track the follow-up work.
